SU2_CFD/include/solvers/CScalarSolver.inl

```diff
@@ -262,6 +287,18 @@ void CScalarSolver<VariableType>::Upwind_Residual(CGeometry* geometry, CSolver**
         if (implicit) Jacobian.UpdateBlocks(iEdge, iPoint, jPoint, residual.jacobian_i, residual.jacobian_j);
       }
 
+      /*--- Applying convective flux correction to negate the effects of flow divergence ---*/
+      if(bounded_scalar){
+        for(iVar=0; iVar<nVar; iVar++){
+          FluxCorrection_i = GetNodes()->GetSolution(iPoint, iVar) * EdgeMassFlux;
+          FluxCorrection_j = GetNodes()->GetSolution(jPoint, iVar) * EdgeMassFlux;
+
+          LinSysRes(iPoint, iVar) -= FluxCorrection_i;
+          LinSysRes(jPoint, iVar) += FluxCorrection_j;
+          Jacobian_i_correction[iVar][iVar] = max(0.0, EdgeMassFlux);;
+          Jacobian_j_correction[iVar][iVar] = min(0.0, EdgeMassFlux);
+        }
+      }if (implicit) Jacobian.UpdateBlocks(iEdge, iPoint, jPoint, Jacobian_i_correction, Jacobian_j_correction);
```

Comment:
I think I had left a comment here :thinking: this implicit update doesn't look right. It looks like you just need to update the diagonals
```
Jacobian.AddVal2Diag(iPoint, -EdgeMassFlux);
Jacobian.AddVal2Diag(jPoint, EdgeMassFlux);
```
Or is this accounting for what the normal upwind scheme already applied?